frontend/src/components/assessments/student-experience/QuestionNumbers.tsx

```diff
@@ -0,0 +1,50 @@
+import React, { useContext } from "react";
+import { GridItem, SimpleGrid } from "@chakra-ui/react";
+
+import StudentContext from "../../../contexts/StudentContext";
+import QuestionNumberTypes from "../../../types/QuestionNumberTypes";
+
+import QuestionNumber from "./QuestionNumber";
+
+const QuestionNumbers = (): React.ReactElement => {
+  const {
+    test,
+    answers,
+    currentQuestionIndex,
+    setCurrentQuestionIndex,
+  } = useContext(StudentContext);
+
+  const isCompletedQuestion = (index: number) => {
+    return answers[index].elements.every(
+      (element) => element.elementAnswers.length !== 0,
+    );
```

Comment:
I feel like there has to be a better way than iterating every question element on each render. Could we maybe cache this value in the top-level `Answer` object and update it in `updatedAnswer`/`replaceAnswer`? (like store `numCompleted` or something and just check if it equals the length)

Reply:
+1 that makes 100000x more sense, thank you!